### PR TITLE
SALTO-1209: Support partial fetch in salesforce

### PR DESCRIPTION
**Pull request by ori-moisis**
Add support for partial fetch in salesforce - adding a `fetch.target` configuration that allows
fetching a specific metadata type.

Also in this commit:
- Fix business hours settings missing fields
- Make reference annotations consistently pick metadata type over custom object

---

Still missing from this PR:
- e2e test in cli
- more unit tests

---
_Release Notes_: 
Salesforce adapter
- Added option for partial fetch using the `salesforce.fetch.target` configuration key

### CODE REVIEW

File: packages/salesforce-adapter/src/filters/global_value_sets.ts [hunk 1]
```diff
@@ -13,57 +13,54 @@
...
     .forEach(f => {
       const valueSetName = getValueSetName(f)
-      if (valueSetName && gvsToRef[valueSetName]) {
-        f.annotations[VALUE_SET_FIELDS.VALUE_SET_NAME] = gvsToRef[valueSetName]
+      if (valueSetName === undefined) {
+        return
+      }
+      const valueSetId = gvsToRef.get(valueSetName)
+      if (valueSetId) {
+        f.annotations[VALUE_SET_FIELDS.VALUE_SET_NAME] = new ReferenceExpression(valueSetId)
```
Comment (netama) [inline packages/salesforce-adapter/src/filters/global_value_sets.ts:44]: nit: is the switch from ref to id (that requires creating a new ref expression here) needed? (mostly asking in case we decide to populate the values in the reference expressions and then we'd need to re-populate it here, if it's actually improving something we can probably worry about it if/when we do that)

File: packages/salesforce-adapter/src/filters/global_value_sets.ts [hunk 2]
```diff
@@ -13,57 +13,54 @@
...
  */
-const filterCreator = (): FilterWith<'onFetch'> => ({
+const filterCreator: FilterCreator = ({ config }): FilterWith<'onFetch'> => ({
   /**
    * @param elements the already fetched elements
    */
   onFetch: async (elements: Element[]): Promise<void> => {
-    const valueSetNameToRef = getValueSetNameToRef(elements)
+    const referenceElements = buildElementsSourceForFetch(elements, config)
+    const valueSetNameToRef = await multiIndex.keyByAsync({
```
Comment (netama) [inline packages/salesforce-adapter/src/filters/global_value_sets.ts:58]: just to make sure I'm not missing anything, the only difference between `keyByAsync` and `buildMultiIndex` is that `keyByAsync` returns just one index? and does it have any advantage over the old implementation in this case?

File: packages/salesforce-adapter/src/filters/layouts.ts [hunk 3]
```diff
@@ -60,28 +58,43 @@ const fixLayoutPath = (
...
+      filter: isCustomObject,
+      key: obj => [apiName(obj)],
+      map: obj => obj.elemID,
     })
+
+    await series(
+      layouts.map(layout => async () => {
+        const [layoutObjName, layoutName] = layoutObjAndName(layout)
+        const layoutObjId = apiNameToCustomObject.get(layoutObjName)
+        const layoutObj = layoutObjId === undefined
```
Comment (netama) [inline packages/salesforce-adapter/src/filters/layouts.ts:86]: same nit as above re reversing the order

File: packages/salesforce-adapter/src/filters/reference_annotations.ts [hunk 4]
```diff
@@ -48,7 +54,7 @@ const convertAnnotationsToReferences = (
   }
 
   elements
-    .filter(isObjectType)
+    .filter(isMetadataTypeOrCustomObject)
```
Comment (netama) [inline packages/salesforce-adapter/src/filters/reference_annotations.ts:57]: just to make sure I'm not missing anything - this is not directly related to this PR, right?

File: packages/salesforce-adapter/src/filters/replace_instance_field_values.ts [hunk 5]
```diff
@@ -47,17 +46,36 @@ const fieldSelectMapping = [
...
-const getApiNameToIdLookup = async (client: SalesforceClient): Promise<Record<string, string>> => (
-  _.mapValues(await getIdsForType(client, CUSTOM_FIELD), toShortId)
-)
+const getRelevantFieldMapping = async (
+  elementsSource: ReadOnlyElementsSource,
+  key: (field: Field) => [string],
+  value: (field: Field) => string,
+): Promise<multiIndex.Index<[string], string>> => {
+  const isReferencedCustomObject = (elem: Element): elem is ObjectType => (
+    isCustomObject(elem)
```
Comment (netama) [inline packages/salesforce-adapter/src/filters/replace_instance_field_values.ts:55]: this check looks like a new addition (compared to https://github.com/salto-io/salto/pull/1902/files#diff-901f9a32f538e632ba58d889d3790999983ed401ca83fba4fd3324f7b35c86b3L97 )?

File: packages/salesforce-adapter/src/filters/replace_instance_field_values.ts [hunk 6]
```diff
@@ -91,41 +112,35 @@ const replaceInstancesValues = (elements: Element[], nameLookUp: Record<string,
...
+      field => [toShortId(getInternalId(field))],
+      apiName,
+    )
     replaceInstancesValues(elements, idToApiNameLookUp)
   },
   preDeploy: async (changes: ReadonlyArray<Change>): Promise<void> => {
-    const apiNameToIdLookup = await getApiNameToIdLookup(client)
+    const apiNameToIdLookup = await getRelevantFieldMapping(
+      config.elementsSource,
+      field => [apiName(field)],
```
Comment (netama) [inline packages/salesforce-adapter/src/filters/replace_instance_field_values.ts:131]: nit: since both `key` and `map` have a similar format, may be worth using named parameters?

File: packages/salesforce-adapter/src/filters/settings_type.ts [hunk 7]
```diff
@@ -73,17 +73,13 @@ const filterCreator: FilterCreator = ({ client, config }) => ({
...
-      elements.filter(isObjectType).map(e => [apiName(e), e])
-    )
-
     const settingsTypeInfos = settingsList.filter(
       info => (config.fetchProfile.metadataQuery)
         .isTypeMatch(getSettingsTypeName(info.fullName))
     )
 
     // Create settings types
+    const knownTypes = new Map()
```
Comment (netama) [inline packages/salesforce-adapter/src/filters/settings_type.ts:82]: can you explain this change? :) does it just look confusing in the diff?

File: packages/salesforce-adapter/src/filters/standard_value_sets.ts [hunk 8]
```diff
@@ -203,14 +206,21 @@ export const makeFilter = (
...
+
+      configChanges = svsInstances.configChanges
+      fetchedSVSInstances = svsInstances.elements
+    }
+
+    const customObjectTypeElements = elements.filter(isCustomObject)
+
+    if (customObjectTypeElements.length > 0) {
+      const svsInstances = fetchedSVSInstances ?? await toArrayAsync(filterAsync(
+        await config.elementsSource.getAll(), isInstanceOfType(STANDARD_VALUE_SET),
```
Comment (netama) [inline packages/salesforce-adapter/src/filters/standard_value_sets.ts:218]: does this mean we never get it from the latest fetched elements? or am I not reading this correctly (likely)?

File: packages/salesforce-adapter/src/filters/standard_value_sets.ts [hunk 9]
```diff
@@ -203,14 +206,21 @@ export const makeFilter = (
...
+      ))
+      updateSVSReferences(customObjectTypeElements, svsInstances)
     }
-    // [GF] No StandardValueSet MetadataType was found.
```
Comment (netama) [inline packages/salesforce-adapter/src/filters/standard_value_sets.ts:209]: not needed anymore?

File: packages/salesforce-adapter/src/transformers/missing_fields.json [hunk 10]
```diff
@@ -2123,5 +2123,66 @@
         "type": "EntitlementProcessMilestoneItem"
       }
     ]
+  },
+  {
```
Comment (netama) [inline packages/salesforce-adapter/src/transformers/missing_fields.json:2127]: if this is not related, maybe do this in a separate PR?

File: packages/salesforce-adapter/specific-cli-options.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
+# Salesforce Specific CLI options
+
+The salesforce adapter supports several options that can be specified in certain salto operations via the `-C` parameter:
+
+| Name                                 | Example value              | Relevant commands
+| -----------------------------------  | -------------------------- | ------------------
+| `salesforce.fetch.target`            | ["CustomObject", "Layout"] | `fetch`
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/specific-cli-options.md:7]: I considered it, but I think for the command line it is not so easy to translate between the table format and the "what do I actually need to write in the command line" format, so I chose to write it in one table for now.

do you think it would be better to have the other table layout?

File: packages/salesforce-adapter/src/filters/custom_objects.ts [hunk 12]
```diff
@@ -533,28 +537,36 @@ const fixDependentInstancesPathAndSetParent = (elements: Element[]): void => {
...
-    .filter(hasCustomObjectParent)
-    .forEach(instance => {
-      const customObj = getDependentCustomObj(instance)
-      if (_.isUndefined(customObj)) {
-        return
-      }
-      setDependingInstancePath(instance, customObj)
-      addObjectParentReference(instance, customObj)
-    })
+  await series(
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/custom_objects.ts:557]: no reason, kind of did it automatically because the old code ran one-by-one
will change

File: packages/salesforce-adapter/src/filters/custom_objects.ts [hunk 13]
```diff
@@ -846,8 +858,11 @@ const filterCreator: FilterCreator = ({ client, config }) => {
       newElements
         .filter(newElem => !elementFullNames.has(id(newElem)))
         .forEach(newElem => elements.push(newElem))
-      fixDependentInstancesPathAndSetParent(elements)
-      removeUnsupportedFields(elements, config[UNSUPPORTED_SYSTEM_FIELDS] ?? [])
+      await fixDependentInstancesPathAndSetParent(
+        elements,
+        buildElementsSourceForFetch(elements, config)
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/custom_objects.ts:863]: I don't think a comment in one usage of the function would be very helpful (and I wouldn't want to add this comment every time the function is used)
would you prefer that I pass in `config.elementsSource` and `config.fetchProfile.isPartialFetch` separately to make it clearer that this is related to partial fetch?

File: packages/salesforce-adapter/src/filters/extra_dependencies.ts [hunk 14]
```diff
@@ -150,46 +124,87 @@ const addGeneratedDependencies = (elem: Element, refElemIDs: ElemID[]): void =>
...
+      if (elemId === undefined) {
+        log.debug(
+          'Element %s:%s (%s) no found, skipping %d dependencies',
+          edge.from.type, edge.from.id, edge.from.name, edge.to.length,
+        )
+        return
+      }
+      const elem = await getFetchedElement(elemId)
+      if (elem === undefined) {
+        log.debug(
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/extra_dependencies.ts:163]: if this element was not fetched, it came from a **read only** element source.
even if we did update an element that came from there (which I think we really really really shouldn't), I don't think it would actually cause it to be updated in the nacl, not sure what would happen, but it probably won't be good

File: packages/salesforce-adapter/specific-cli-options.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (netama) [reply packages/salesforce-adapter/specific-cli-options.md:7]: I think if we also keep the examples you added below it can still be clear, but mostly asked because of the inconsistency, I don't have a strong opinion on that. up to you (and maybe check with Tomer?)

File: packages/salesforce-adapter/src/filters/custom_objects.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (netama) [reply packages/salesforce-adapter/src/filters/custom_objects.ts:863]: no, I guess people will get used to it then :)

File: packages/salesforce-adapter/src/filters/extra_dependencies.ts [hunk 15]
```diff
@@ -150,46 +124,87 @@ const addGeneratedDependencies = (elem: Element, refElemIDs: ElemID[]): void =>
...
  * Add references using the tooling API.
  */
-const creator: FilterCreator = ({ client }) => ({
+const creator: FilterCreator = ({ client, config }) => ({
   onFetch: async (elements: Element[]) => {
     const groupedDeps = await getDependencies(client)
-    const elemLookup = generateElemLookup(elements)
-    const customObjectLookup = generateCustomObjectLookup(elements)
-    await addExtraReferences(groupedDeps, elemLookup, customObjectLookup)
+    const fetchedElements = buildElementsSourceFromElements(elements)
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/extra_dependencies.ts:189]: AFAICT the "cost" of building the element source currently is just in the mapping it creates between the elem IDs and top level elements (so, one iteration over the elements and one map)
since these are only top level elements I guess it would be a mapping of ~100k items on large accounts? shouldn't take more than a second or two (not a big deal for a fetch operation)

in this filter we used to make two iterations over the elements to create two mappings.
now on a regular fetch we make three iterations (one for each element source and one for building the two original mappings).
so my estimate would be that it would at most a couple of seconds to a filter where the CPU part would take a few seconds anyway and where CPU time is probably negligible compared to the time the API call takes.

generally (not just in this filter), we create the element sources in filters that used to create a mapping of elements anyway, so at most we'd be doubling the time it takes to create the index of elements which I hope is not the dominant factor in most filters (and I generally believe that it would be measured in seconds even for large accounts so it should not make a noticeable impact of the total fetch time)

of course this is mostly theoretical analysis, so it is very possible I missed something.
if you have something to suggest let me know

File: packages/salesforce-adapter/src/filters/extra_dependencies.ts [hunk 16]
```diff
@@ -150,46 +124,87 @@ const addGeneratedDependencies = (elem: Element, refElemIDs: ElemID[]): void =>
...
+        key: elem => [metadataType(elem), getInternalId(elem)],
+        map: elem => elem.elemID,
+      })
+      .addIndex({
+        name: 'customObjectLookup',
+        filter: isCustomObject,
+        key: elem => [apiName(elem)],
+        map: elem => elem.elemID,
+      })
+      .process(flatMapAsync(await allElements.getAll(), extractFlatCustomObjectFields))
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/extra_dependencies.ts:205]: right now I am not sure which one would win. the new code has more compute operations, but less memory operations (loading the elements from RAM into CPU working memory once instead of twice)

I do think that this approach will be much better than making two separate iterations after lazy element intg is merged because then loading elements into memory will likely be much slower

File: packages/salesforce-adapter/src/filters/field_references.ts [hunk 17]
```diff
@@ -225,37 +226,35 @@ const replaceReferenceValues = (
...
-): void => {
+): Promise<void> => {
   const resolverFinder = generateReferenceResolverFinder(defs)
-  const elemLookup = groupByMetadataTypeAndApiName(elements)
+
+  const elementsWithFields = flatMapAsync(
+    await referenceElements.getAll(),
+    extractFlatCustomObjectFields,
+  )
+  // TODO - when transformValues becomes async the first index can be to elemID and not the whole
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/field_references.ts:240]: part of lazy element intg

File: packages/salesforce-adapter/src/filters/field_references.ts [hunk 18]
```diff
@@ -225,37 +226,35 @@ const replaceReferenceValues = (
...
+  // element and we can use the element source directly instead of creating the second index
+  const { elemLookup, elemByElemID } = await multiIndex.buildMultiIndex<Element>()
+    .addIndex({
+      name: 'elemLookup',
+      filter: hasExternalId,
+      key: elem => [metadataType(elem), apiName(elem)],
+      map: elem => elem,
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/field_references.ts:247]: no, missed that one when I made `map` optional
thanks

File: packages/salesforce-adapter/src/filters/field_references.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (netama) [reply packages/salesforce-adapter/src/filters/field_references.ts:240]: :+1: if this TODO is intended to be merged then maybe mention one of the tickets by id?

File: packages/salesforce-adapter/src/filters/foreign_key_references.ts [hunk 19]
```diff
@@ -43,9 +45,7 @@ const resolveReferences = (
 
     const refTarget = makeArray(field.annotations[FOREIGN_KEY_DOMAIN])
       .filter(isReferenceExpression)
-      .map((ref: ReferenceExpression) => apiNameToElemIDs[value]?.[
-        ref.elemId.typeName
-      ])
+      .map(ref => apiNameToElemIDs.get(ref.elemId.typeName, value))
```
[the remark this answers is not in the document]
Comment (netama) [reply packages/salesforce-adapter/src/filters/foreign_key_references.ts:48]: it didn't mention it by name because it was a secondary index and the name would be too long, but the reason `apiName` specifically was mentioned was that it was the first index, which it no longer is

File: packages/salesforce-adapter/src/filters/global_value_sets.ts (hunk 1 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/global_value_sets.ts:44]: the old code created references as well, not sure I understand the question

File: packages/salesforce-adapter/src/filters/foreign_key_references.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/foreign_key_references.ts:48]: `externalIDToElemIDs`?

File: packages/salesforce-adapter/src/filters/replace_instance_field_values.ts (hunk 5 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/replace_instance_field_values.ts:55]: changed from `isObjectType` because only fields of custom objects can be relevant - only fields of custom objects have IDs.
I can change it back to `isObjectType` if you want, because the second condition would drop all the non-custom objects anyway and we have an explicit `hasInternalId` filter on the index later.

let me know if you want me to change it

Comment (netama) [reply packages/salesforce-adapter/src/filters/replace_instance_field_values.ts:55]: same as above - if we didn't lose any cases in the transition (like something that was using the hard-coded internal-id replacement for standard fields) then it's probably ok, but wanted to make sure we're not dropping references that we were able to generate before

File: packages/salesforce-adapter/src/filters/settings_type.ts (hunk 7 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/settings_type.ts:82]: oops, this should have been part of PR #1854
I must have missed it then.
basically given the change that was made in that PR, it is no longer necessary to check the pre-existing types when describing the settings types.

File: packages/salesforce-adapter/src/filters/standard_value_sets.ts (hunk 8 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/standard_value_sets.ts:218]: this whole thing is after `fetchedSVSInstances ??` , so this will only run in case we didn't fetch the standard value sets.
I can see why this is easy to miss, I wanted to extract this to a function so the `??` would be in one line but I couldn't think of a name that would fit in one line (the name + `config.elementsSource` always ended up being too long and then I'd have to break the line and it would end up looking like this).

one thing I could do is not use `??` here and instead write it as
```typescript
      const svsInstances = fetchedSVSInstances !== undefined
        ? fetchedSVSInstances
        : await toArrayAsync(
          filterAsync(await config.elementsSource.getAll(), isInstanceOfType(STANDARD_VALUE_SET))
        )
```
maybe that would be clearer?

File: packages/salesforce-adapter/src/filters/utils.ts [hunk 20]
```diff
@@ -196,18 +194,26 @@ export const setInternalId = (elem: Element, val: string): void => {
   }
 }
 
-// ApiName -> MetadataType -> ElemID
-export type ApiNameMapping = Record<string, Record<string, ElemID>>
+export const hasInternalId = (elem: Element): boolean => (
+  getInternalId(elem) !== undefined && getInternalId(elem) !== ''
+)
+
+export const hasExternalId = (elem: Element): boolean => (
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/utils.ts:201]: how would you call it?

File: packages/salesforce-adapter/src/transformers/missing_fields.json (hunk 10 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/transformers/missing_fields.json:2127]: again, should have been a part of #1854
should I extract these two parts to a separate PR?

File: packages/salesforce-adapter/src/filters/reference_annotations.ts (hunk 4 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/reference_annotations.ts:57]: we did not lose anything.
as I tried to write above, this is filtering out types that cannot have references because they did not come from salesforce.

File: packages/salesforce-adapter/src/filters/replace_instance_field_values.ts (hunk 5 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/replace_instance_field_values.ts:55]: we're not dropping references that we were able to generate before

File: packages/salesforce-adapter/src/filters/standard_value_sets.ts (hunk 8 above)
Comment (netama) [reply packages/salesforce-adapter/src/filters/standard_value_sets.ts:218]: I knew I was missing something... :+1: for the rewrite, or maybe also just with parens 
```
const svsInstances = (
  fetchedSVSInstances
  ?? await toArrayAsync(filterAsync(await config.elementsSource.getAll(), isInstanceOfType(STANDARD_VALUE_SET),
)
```
but you can also just keep it this way

File: packages/salesforce-adapter/src/filters/utils.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (netama) [reply packages/salesforce-adapter/src/filters/utils.ts:201]: I'd keep the name and add a comment above it explaining why both are needed (I'm guessing it's related to the salesforce requirements because you mentioned something about it last week, but otherwise I'd just assume it's something salto-specific that I don't understand and I think that's what comments are for)

[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/utils.ts:201]: I think we are misunderstanding each other.
I can change this to check only the `apiName(elem) !== undefined` if you think that would be clearer (I think everything that has an `apiName` also has a `metadataType`)

File: packages/salesforce-adapter/test/filters/extra_dependencies.test.ts [hunk 21]
```diff
@@ -108,71 +122,110 @@ describe('Internal IDs filter', () => {
...
+          MetadataComponentId: 'inst1 id',
+          MetadataComponentName: 'n1',
+          RefMetadataComponentType: 'meta2',
+          RefMetadataComponentId: 'inst3 id',
+          RefMetadataComponentName: 'n2',
+        },
+        {
+          MetadataComponentType: 'meta2',
+          MetadataComponentId: 'inst3 id',
+          MetadataComponentName: 'n1',
```
Comment (netama) [inline packages/salesforce-adapter/test/filters/extra_dependencies.test.ts:180]: nit (not sure if new or my fault... but visible here): probably better to keep the names consistent with the type+id combos (it can just be the same as the id, but this and the previous item had the names swapped)

**Review by netama — APPROVED** [review]
LGTM (with some small nits) after e2e

File: packages/salesforce-adapter/test/filters/extra_dependencies.test.ts (hunk 21 above)
Comment (netama) [reply packages/salesforce-adapter/test/filters/extra_dependencies.test.ts:180]: in reality `Id` is the internal id and `Name` is usually something like the api name (and we don't use the name for resolving which is why it doesn't actually matter). I think in the test it's clearer if the id is not random characters like the real internal id, so probably easiest to just have uppercase/lowercase or something like that? or identical name and id should also be ok. thanks!
